### PR TITLE
Add DefaultQueryParamDecoderMatcher class

### PR DESCRIPTION
**Pull request by sbly**
Hello! First-time contributor here. This PR adds a `DefaultQueryParamDecoderMatcher`, which, as the name suggests, accepts a default value for a query parameter. I followed the pattern for the other matchers. Let me know if there's anything I missed!

### CODE REVIEW

**Review by armanbilge — APPROVED** [review]
LGTM, nice work following the existing patterns :)

**Comment by sbly** [comment]
@armanbilge Thanks! Are you able to merge? It seems one CI run was unsuccessful, but looking at the error, it doesn't seem to be related to my PR. Maybe it was just a fluke?

**Comment by armanbilge** [comment]
Yes, don't worry about CI, just a flake :) I was hoping to get one more pair of eyes before merging, let's give someone a nudge 😉

**Comment by armanbilge** [comment]
Oh actually, @rossabaker will cherry-pick this into 0.22 at the next opportunity. Unless you want to retarget against series/0.22?

File: dsl/src/main/scala/org/http4s/dsl/impl/Path.scala
Comment (bplommer) [inline dsl/src/main/scala/org/http4s/dsl/impl/Path.scala:330]: I think this should return `Some` rather than `Option`?

Comment (armanbilge) [reply dsl/src/main/scala/org/http4s/dsl/impl/Path.scala:330]: Are you sure? What if `params.get(name)` is `None`?

File: dsl/src/main/scala/org/http4s/dsl/impl/Path.scala
Comment (bplommer) [inline dsl/src/main/scala/org/http4s/dsl/impl/Path.scala:329]: Would `QueryParamDecoderMatcherWithDefault` be a better name? We're not deciding a default query peram.

**Comment by bplommer** [comment]
I like the idea, but is there a more compositional way we can do this so that defaults can be applied to arbitrary optional matchers? The answer may be no, given the requirement for stable identifiers.

File: dsl/src/main/scala/org/http4s/dsl/impl/Path.scala
Comment (bplommer) [reply dsl/src/main/scala/org/http4s/dsl/impl/Path.scala:330]: Isn't that what the default is for?

Sounds like this should have a doc to spell out the semantics :)

Comment (armanbilge) [reply dsl/src/main/scala/org/http4s/dsl/impl/Path.scala:330]: Haha yes I derped there 😆 good point.

Comment (bplommer) [reply dsl/src/main/scala/org/http4s/dsl/impl/Path.scala:330]: Does it also use the default if there's a value but it fails to parse? (I'm having one of those days where my code-reading comprehension is extremely minimal.) I think that should be documented either way.

Comment (armanbilge) [reply dsl/src/main/scala/org/http4s/dsl/impl/Path.scala:330]: But I remember looking at the tests and they seemed to do the right thing. So now I am confused.

Comment (bplommer) [reply dsl/src/main/scala/org/http4s/dsl/impl/Path.scala:330]: If it still compiles after changing the `Option` to `Some` then the behaviour is probably correct.

**Comment by armanbilge** [comment]
> is there a more compositional way we can do this so that defaults can be applied to arbitrary optional matchers?

This sounds cool, but what would it compose with?

**Comment by bplommer** [comment]
Possibly a trait `DefaultParamValue` that can be mixed into other matcher classes?

**Comment by bplommer** [comment]
> > is there a more compositional way we can do this so that defaults can be applied to arbitrary optional matchers?
> 
> 
> 
> This sounds cool, but what would it compose with?

An underlying non-total matcher.

Oh, did my other comment only just send?

**Comment by sbly** [comment]
@bplommer Initially I tried to extend `OptionalQueryParamDecoderMatcher` and override the `unapply` method, but this didn't work because they have different return types: `Option[Option[T]]` vs `Option[T]`. I don't think we can extend `QueryParamDecoderMatcher` either because at that point we've already swallowed the error. Or did you mean something else?

File: dsl/src/main/scala/org/http4s/dsl/impl/Path.scala
Comment (sbly) [reply dsl/src/main/scala/org/http4s/dsl/impl/Path.scala:330]: It's `Option` because we still return `None` in the case that the parameter is present but with an incorrect format, eg as in one of the test cases. However I can change this behavior to return the default in that case if you prefer that

Comment (sbly) [reply dsl/src/main/scala/org/http4s/dsl/impl/Path.scala:330]: Haha no problem, I'll add some docs

Comment (bplommer) [reply dsl/src/main/scala/org/http4s/dsl/impl/Path.scala:333]: I was going to say it should be changed, but then I saw the optional decoder returns `None` on the outer option in the case of a decoding failure. I disagree with that behaviour but I think for consistency we should do the same here. I'd like to reconsider that before 1.0 but I don't think we should hold this PR up on account of that.

**Review by bplommer — APPROVED** [review]

File: dsl/src/main/scala/org/http4s/dsl/impl/Path.scala
Comment (armanbilge) [reply dsl/src/main/scala/org/http4s/dsl/impl/Path.scala:333]: Interesting, look forward to hear your thoughts on that :)

**Review by rossabaker — COMMENTED** [review]
Thanks, @sbly.  As mentioned above, I'll cherry-pick this back so it can be used starting in 0.22.

**Pull request closed without merging**
